### PR TITLE
Drop older Python versions

### LINKED ISSUE #819: Drop Python 2.7 support
2.7 support in eventlet has been deprecated since December 2020, so 3 years at time of opening this issue, and supporting it seems like a bottleneck for other improvements. Compatibility across 13 years of Python versions seems... tricky.

As context CPython stopped security updates for 2.7 in January 2020.

I can submit a PR if this is acceptable.

### LINKED ISSUE #822: Drop Python 3.5, 3.6, 3.7
None of these are getting security updates anymore, so there's no point in supporting them.

### PR DESCRIPTION
**Pull request by itamarst**
Fixes #819
Fixes #822 

Tests won't pass for now, but should at least run as before, just fewer builders.

### CODE REVIEW

**Comment by 4383** [comment]
I'd suggest to squash the commits

**Comment by 4383** [comment]
at least those related to version removing.

**Comment by jstasiak** [comment]
FYI, regarding commit squashing: we've been using squash merges for some time now in this project. Depending on how you want to organize your work the PR-local commit noise may not necessarily matter much.

**Comment by 4383** [comment]
> FYI, regarding commit squashing: we've been using squash merges for some time now in this project. Depending on how you want to organize your work the PR-local commit noise may not necessarily matter much.

You are right, no problem, let's use squash merge.

**Comment by damani42** [comment]
Yes I agree about the squash merge actually, it's a bit easier for the contributors.

File: tox.ini [hunk 1]
```diff
@@ -20,8 +20,8 @@ envlist =
     pep8
     py38-openssl
     py39-dnspython1
-    py{27,35,36,py2,py3}-epolls
-    py{37,38,39,310}-{selects,poll,epolls}
+    pypy3-epolls
+    py{38,39,310}-{selects,poll,epolls}
```
Comment (4383) [inline tox.ini:24]: Not sure why the deps of below `testenv` manage Python 3.11 and why the above `envlist`, apparently, not. As the semantic of these changes is to drop EOLed Python releases, this patch LGTM.

**Review by 4383 — APPROVED** [review]
LGTM

**Comment by 4383** [comment]
@temoto, @jstasiak: Even if I'm now an eventlet orga team member, I'm not able to merge patches from the eventlet repo. I think you need to tweak some specific rights of this repo.

However, in a first time I'll only focus on easy patches like this one, to become more familiar with the eventlet repo and avoid mistake.

**Comment by jstasiak** [comment]
> @temoto, @jstasiak: Even if I'm now an eventlet orga team member, I'm not able to merge patches from the eventlet repo.

I think that's because you were in maintainers but not in develoepers. I just added you to the developers team, try now.

**Comment by 4383** [comment]
> > @temoto, @jstasiak: Even if I'm now an eventlet orga team member, I'm not able to merge patches from the eventlet repo.
> 
> I think that's because you were in maintainers but not in develoepers. I just added you to the developers team, try now.

Looks better, thanks.

@itamarst: However, we should notice that the merge is still disabled due to the required tests that are failing, so what do you think of my [development branch proposal made there](https://github.com/eventlet/eventlet/issues/828#issuecomment-1855526862)?

Using this development branch we could temporarly disable failing tests, then redirect this pull request and your other one (https://github.com/eventlet/eventlet/pull/823) against the development branch rather than against the master branch, then, merge them, and finally re-enabling all the tests.

That would avoid disabling the tests, on all the master pull requests. Even if they tests are not functional, they protect us from merging patches accidentally. Patches merged on `development` could be later cherry-picked to `master`.

**Comment by 4383** [comment]
@jstasiak: I think the same admin action should be done for all the persons recently added. Big thanks for your time and for your help!

**Comment by jstasiak** [comment]
Everyone else was already in the developers team so I don't think anything to do there. Let me know if it turns out otherwise though.

FYI for your convenience I temporarily disabled the requirement for PRs to have green builds to be merged. Let me know if you want it undone (and when).

**Comment by 4383** [comment]
> Everyone else was already in the developers team so I don't think anything to do there. Let me know if it turns out otherwise though.
> 
> FYI for your convenience I temporarily disabled the requirement for PRs to have green builds to be merged. Let me know if you want it undone (and when).

Great, thanks

**Comment by 4383** [comment]
@itamarst: which pull request are you desire to be merged first?:
- this one
- https://github.com/eventlet/eventlet/pull/823

**Comment by itamarst** [comment]
My plan is:

1. Merge this one.
2. Split the "green CI" into two, with one only covering 3.8-3.10, since 3.11 is the most significant change and 3.11 is moreover not even in current CI config.
3. Do a third PR that adds 3.11 support so it's easier to discuss in isolation.

**Comment by tipabu** [comment]
I took a look at recent download stats from https://pypistats.org/packages/eventlet (and used [its API](https://pypistats.org/api/packages/eventlet/python_minor) to get raw stats so I could compute monthly aggregates) -- I worry that dropping 3.7 may have been a little hasty. While 3.7 downloads have been trending down for months, we've only just this month begun getting more downloads from 3.7 than 3.11; 3.7 accounts for roughly one in eight downloads.

| Python<br>Version | July | August | September | October | November |
| ---: | ---: | ---: | ---: | ---: | ---: |
| 2.7 | 34613 | 34128 | 31711 | 32852 | 36821 |
| 3.4 | 91 | 101 | 39 | 42 | 49 |
| 3.5 | 1035 | 534 | 655 | 640 | 758 |
| 3.6 | 61010 | 52010 | 45942 | 50901 | 53917 |
| 3.7 | 335613 | 303073 | 255596 | 220816 | 194961 |
| 3.8 | 311678 | 244907 | 277575 | 349432 | 341839 |
| 3.9 | 180688 | 223139 | 194558 | 175212 | 172666 |
| 3.10 | 519239 | 465097 | 294619 | 263098 | 242695 |
| 3.11 | 69017 | 75176 | 69583 | 77139 | 184093 |
| 3.12 | 84 | 269 | 317 | 4062 | 5773 |

Honestly, I'm not sure _why_ there would be so many 3.7 downloads; my recollection was that Ubuntu LTS releases used 2.7, 3.8, 3.10, while RHEL/CentOS hit 2.7, 3.6, 3.9... DistroWatch gives me a hint that it may be [Debian Buster](https://distrowatch.com/table.php?distribution=debian)? If so, at least their long term support would end come the middle of next year...

The question seems particularly important, though, given the premise of [eventlet as critical to whole ecosystems of projects](https://github.com/eventlet/eventlet/issues/824) and our need to [address security issues](https://github.com/eventlet/eventlet/pull/826).

**Comment by 4383** [comment]
@thomasgoirand o/ Any opinion about to the previous comment?

**Comment by itamarst** [comment]
I opened a new issue for the 3.7 discussion: https://github.com/eventlet/eventlet/issues/846

**Pull request merged**